File: app/src/main/java/com/xephorium/crystalnote/ui/selection/SelectionActivity.java
package com.xephorium.crystalnote.ui.selection;

import android.os.Bundle;
import android.widget.Toast;

import com.xephorium.crystalnote.R;
import com.xephorium.crystalnote.data.model.Note;
import com.xephorium.crystalnote.ui.base.ToolbarActivity;
import com.xephorium.crystalnote.ui.custom.NoteListView;
import com.xephorium.crystalnote.ui.custom.NoteToolbar;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class SelectionActivity extends ToolbarActivity implements SelectionView {

    @BindView(R.id.selection_note_list)
    NoteListView noteListView;

    SelectionPresenter presenter;

    @OnClick(R.id.selection_action_button)
    public void handleActionButtonClick() {
        presenter.handleActionButtonClick();
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setActivityContent(R.layout.selection_activity_layout);
        ButterKnife.bind(this);

        presenter = new SelectionPresenter(this);
        noteListView.setNoteListViewListener(getSelectionNoteListViewListener());
        setupSelectionToolbar();
    }

    @Override
    protected void onResume() {
        super.onResume();
        presenter.attachView(this);
        presenter.refreshNoteList();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        presenter.detachView();
    }

    @Override
    public void populateNoteList(List<Note> notes) {
        noteListView.populateNoteList(notes);
    }

    @Override
    public void closeSelectionActivity() {
        this.finish();
    }

    private void setupSelectionToolbar() {
        getToolbar().setEditMode(false);
        getToolbar().setTitle(R.string.selection_title);
        getToolbar().setLeftButtonImage(R.drawable.icon_back);
        getToolbar().setRightButtonImage(R.drawable.icon_search);
        getToolbar().setNoteToolbarListener(getSelectionNoteToolbarListener());
    }

    private NoteToolbar.NoteToolbarListener getSelectionNoteToolbarListener() {
        return new NoteToolbar.NoteToolbarListener() {
            @Override
            public void onLeftButtonClick() {
                presenter.handleToolbarBackClick();
            }

            @Override
            public void onRightButtonClick() {
                Toast.makeText(getApplicationContext(), "Search Notes", Toast.LENGTH_SHORT).show();
            }
        };
    }

    private NoteListView.NoteListViewListener getSelectionNoteListViewListener() {
        return new NoteListView.NoteListViewListener() {
            @Override
            public void onNoteClick(Note note) {
                presenter.handleNoteClick(note);
            }

            @Override
            public boolean onNoteLongClick(Note note) {
                return presenter.handleNoteLongClick(note);
            }

            @Override
            public void onNoteListRefresh() {
                presenter.refreshNoteList();
            }
        };
    }
}

File: app/src/main/java/com/xephorium/crystalnote/ui/selection/SelectionActivity.kt
package com.xephorium.crystalnote.ui.selection

import android.content.Intent
import android.os.Bundle
import com.xephorium.crystalnote.R
import com.xephorium.crystalnote.data.NoteRepository
import com.xephorium.crystalnote.data.SharedPreferencesRepository
import com.xephorium.crystalnote.data.model.Note
import com.xephorium.crystalnote.ui.IntentLibrary
import com.xephorium.crystalnote.ui.base.ToolbarActivity
import com.xephorium.crystalnote.ui.update.UpdateActivity
import com.xephorium.crystalnote.ui.custom.NoteListView
import com.xephorium.crystalnote.ui.custom.NoteToolbar

import kotlinx.android.synthetic.main.selection_activity_layout.*
import kotlinx.android.synthetic.main.toolbar_activity_layout.*

class SelectionActivity : ToolbarActivity(), SelectionContract.View {


    /*--- Variable Declarations ---*/

    lateinit var presenter: SelectionPresenter


    /*--- Lifecycle Methods ---*/

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setActivityContent(R.layout.selection_activity_layout)

        presenter = SelectionPresenter()
        presenter.noteRepository = NoteRepository(this)
        presenter.sharedPreferencesRepository = SharedPreferencesRepository(this)

        setupToolbar()
        setupClickListeners()
    }

    override fun onResume() {
        super.onResume()
        presenter.attachView(this)
    }

    public override fun onDestroy() {
        super.onDestroy()
        presenter.detachView()
    }


    /*--- View Manipulation Methods ---*/

    override fun populateNoteList(notes: List<Note>) {
        selection_note_list.populateNoteList(notes)
    }

    override fun refreshWidget() {
        sendBroadcast(Intent().also { it.action = IntentLibrary.UPDATE_NOTE_INTENT })
    }

    override fun navigateBack() {
        finish()
    }

    override fun navigateToNewNote() {
        val intent = Intent(this, UpdateActivity::class.java)
        intent.action = IntentLibrary.CREATE_NOTE
        startActivity(intent)
    }


    /*--- Private Setup Methods ---*/

    private fun setupToolbar() {
        toolbar.isEditMode = false
        toolbar.setTitle(R.string.selection_title)
        toolbar.setLeftButtonImage(R.drawable.icon_back)
        toolbar.setNoteToolbarListener(object : NoteToolbar.NoteToolbarListener {
            override fun onLeftButtonClick() = presenter.handleToolbarBackClick()
            override fun onRightButtonClick() = Unit
            override fun onTextChange(text: String) = Unit
        })
    }

    private fun setupClickListeners() {
        selection_action_button.setOnClickListener { presenter.handleNewNoteButtonClick() }
        selection_note_list.noteListViewListener = object : NoteListView.NoteListViewListener {
            override fun onNoteClick(note: Note) = presenter.handleNoteClick(note)
            override fun onNoteLongClick(note: Note) = presenter.handleNoteLongClick(note)
            override fun onNoteListRefresh() = presenter.handleNoteListRefresh()
        }
    }
}
